examples/nft-pawn-shop/index.rsh

```diff
@@ -30,15 +34,24 @@ export const main = Reach.App(() => {
   commit();
   
   B.interact.loanMade();
-  B.pay(loanAmt + interest)
-   .timeout(relativeTime(deadline), () => {
-      L.publish();
-      transfer(collateral).to(L);
-      commit();
-      exit();
-    });
-  transfer(collateral).to(B);
-  transfer(loanAmt + interest).to(L);
+
+  try {
+    const [[], k] = call(LA.payBack)
+      .pay(() => loanAmt + interest)
+      .assume(() => check(this == B))
+      .throwTimeout(relativeTime(deadline), null);
+    check(this == B);
```

Comment:
Maybe the same argument applies here... we should be okay with someone else paying B's loan :)

Reply:
And if we do that, then we could do a weird thing and not use any APIs, but just have four different participants... kind of a weird, but maybe we discovered a new way to write Reach programs.

Reply:
Even if we did want to enforce that it was B, you could make a new participant and set them to be B

Reply:
Would doing it that way not require someone to be attached the whole time still? I don't understand much about participant internals and what happens when you set them.

Reply:
First, to be clear about what I'm talking about... I'm suggesting that there are four participants:

1. Borrower
2. Lender
3. Repayer
4. Claimant

The program is...

1. Borrower deposits collateral
2. Lender gives loan to Borrower
3. Repayer repays loan and gives collateral and interest to Borrower
4. OR, Claimant waits until deadline and gives collateral to Lender

If Repayer and Claimant were respectively Borrower and Lender, then they'd have to run the program forever, because there's no way to "reattach". "To run the Borrower" would mean "launch the program, deposit collateral, wait, do something". That means that if the Borrower accidentally closed the window, they couldn't get back to the "do something" step.

The way I'm proposing... the Jon would run the program first as Borrower and then close it. Later, they would open it and run the program as Repayer (which would observe the previous actions of the Borrower and Lender) and finish it. Sara could do the same thing as Lender and then Claimant. OR, someone else could run the second parts if they wanted to pay for someone else's lunch.

Reply:
I like this alternative style for this program. it makes the code shorter and cleaner. Though, if there were more publishes to do, we would have to add an interact function and throws. Would it make sense to have a wrapper for that functionality (throwing and catching for intentional disconnects) in the js stdlib?

Reply:
I'm not sure how to do that wrapper in a way that is shorter than just writing out the normal JS pattern... maybe

```
withDisconnect(async (disconnect) => {
 ....
 ready: {
  disconnect();
 }
});
```

where `withDisconnect` is basically `try/catch` and `disconnect` is throw?

Reply:
That's pretty much what I was thinking. I would do it like
```js
stdlib.disconnect = () => throw "reach disconnect";
stdlib.withDisconnect = async (prom) => {
  try { await prom() }
  catch (e) { if (e !== "reach disconnect") throw e;
}

await stdlib.withDisconnect(backend.Alice(ctcAlice, { ready: stdlib.disconnect });
```

Reply:
Seems like a common pattern in testing. Not sure if it would be a common pattern in real usage.

Reply:
Make a Jira card for it, BUT... don't use that string... you should make an object and use object equality to get a ptr to it